File: Editor/src/main/java/org/turbomedia/turboedit/editor/components/MenuBar.java
package org.turbomedia.turboedit.editor.components;

import org.turbomedia.turboedit.editor.ProjectManager;
import org.turbomedia.turboedit.editor.windows.AboutWindow;
import org.turbomedia.turboedit.editor.windows.NewProjectWindow;
import org.turbomedia.turboedit.editor.windows.PreferencesWindow;
import org.turbomedia.turboedit.editor.windows.RenderQueueWindow;
import javafx.scene.control.Menu;
import javafx.scene.control.MenuItem;
import javafx.stage.FileChooser;
import javafx.stage.Stage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.awt.*;
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;

import static org.turbomedia.turboedit.editor.Editor.TITLE;
import static org.turbomedia.turboedit.editor.misc.Locale.GetText;

public class MenuBar extends javafx.scene.control.MenuBar {
    private final Logger logger = LoggerFactory.getLogger(getClass());

    public MenuBar(Stage stage) {
        getMenus().add(getMenuFile(stage));
        getMenus().add(getMenuEdit());

        getMenus().add(getMenuHelp());
    }

    private Menu getMenuFile(Stage stage) {
        var menu = new Menu(GetText("menubar.file"));

        var itemNewProject = new MenuItem(GetText("menubar.file.new_project"));

        itemNewProject.setOnAction(event -> new NewProjectWindow(stage));

        menu.getItems().add(itemNewProject);

        var itemLoadProject = new MenuItem(GetText("menubar.file.load_project"));
        itemLoadProject.setOnAction(event -> {
            var chooser = new FileChooser();

            chooser.getExtensionFilters().addAll(
                    new FileChooser.ExtensionFilter(GetText("file_chooser.ext.tvp"), "*.tvp")
            );

            chooser.setTitle(TITLE + " - " + GetText("title.open_project"));
            var file = chooser.showOpenDialog(stage);

            try {
                ProjectManager.LoadProject(file, stage);
            } catch (IOException e) {
                logger.error("{} = {}", e.getClass().getName(), e.getMessage());
            }
        });

        menu.getItems().add(itemLoadProject);

        var itemPreferences = new MenuItem(GetText("menubar.file.preferences"));
        itemPreferences.setOnAction(event -> new PreferencesWindow());
        menu.getItems().add(itemPreferences);

        var renderQueueMenu = new MenuItem(GetText("menubar.file.render_queue"));
        renderQueueMenu.setOnAction(event -> new RenderQueueWindow());
        menu.getItems().add(renderQueueMenu);

        var itemQuit = new MenuItem(GetText("menubar.file.quit"));
        itemQuit.setOnAction(event -> System.exit(0));

        menu.getItems().add(itemQuit);

        return menu;
    }

    private Menu getMenuEdit() {
        var menu = new Menu(GetText("menubar.edit"));

        return menu;
    }

    private Menu getMenuHelp() {
        var menu = new Menu(GetText("menubar.help"));

        var itemAbout = new MenuItem(GetText("menubar.help.about", TITLE));
        itemAbout.setOnAction(event -> new AboutWindow());

        menu.getItems().add(itemAbout);

        var itemGithub = new MenuItem(GetText("menubar.help.repo"));
        itemGithub.setOnAction(event -> {
            if (Desktop.isDesktopSupported() && Desktop.getDesktop().isSupported(Desktop.Action.BROWSE)) {
                try {
                    Desktop.getDesktop().browse(new URI("https://github.com/Turboman3000/TurboEdit"));
                } catch (IOException | URISyntaxException e) {
                    logger.error("{} = {}", e.getClass().getName(), e.getMessage());
                }
            }
        });
        menu.getItems().add(itemGithub);

        return menu;
    }
}

File: Editor/src/main/kotlin/turboedit/editor/components/MenuBar.kt
package turboedit.editor.components

import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.scene.control.Menu
import javafx.scene.control.MenuBar
import javafx.scene.control.MenuItem
import javafx.stage.FileChooser
import javafx.stage.Stage
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import turboedit.editor.Editor
import turboedit.editor.ProjectManager.loadProject
import turboedit.editor.misc.Locale
import turboedit.editor.misc.Locale.Companion.getText
import turboedit.editor.windows.AboutWindow
import turboedit.editor.windows.NewProjectWindow
import turboedit.editor.windows.RenderQueueWindow
import turboedit.editor.windows.preferences.PreferencesWindow
import java.awt.Desktop
import java.io.IOException
import java.net.URI
import java.net.URISyntaxException
import kotlin.system.exitProcess

class MenuBar(stage: Stage?) : MenuBar() {
    private val logger: Logger = LoggerFactory.getLogger(javaClass)

    init {
        menus.add(getMenuFile(stage))
        menus.add(this.menuEdit)

        menus.add(this.menuHelp)
    }

    private fun getMenuFile(stage: Stage?): Menu {
        val menu = Menu(Locale.getText("menubar.file"))

        val itemNewProject = MenuItem(Locale.getText("menubar.file.new_project"))
        itemNewProject.onAction = EventHandler { event: ActionEvent? -> NewProjectWindow(stage) }
        menu.items.add(itemNewProject)

        val itemLoadProject = MenuItem(Locale.getText("menubar.file.load_project"))
        itemLoadProject.onAction = EventHandler { event: ActionEvent? ->
            val chooser = FileChooser()

            chooser.title = Editor.Companion.TITLE + " - " + Locale.getText("title.open_project")
            chooser.extensionFilters.addAll(
                FileChooser.ExtensionFilter(Locale.getText("file_chooser.ext.tvp"), "*.tvp")
            )

            val file = chooser.showOpenDialog(stage)

            if (file == null) {
                return@EventHandler
            }

            try {
                loadProject(file)
            } catch (e: IOException) {
                logger.error("{} = {}", e.javaClass.getName(), e.message)
            }
        }

        menu.items.add(itemLoadProject)

        val itemPreferences = MenuItem(Locale.getText("menubar.file.preferences"))
        itemPreferences.onAction = EventHandler { event: ActionEvent? -> PreferencesWindow() }
        menu.items.add(itemPreferences)

        val renderQueueMenu = MenuItem(Locale.getText("menubar.file.render_queue"))
        renderQueueMenu.onAction = EventHandler { event: ActionEvent? -> RenderQueueWindow() }
        menu.items.add(renderQueueMenu)

        val itemQuit = MenuItem(Locale.getText("menubar.file.quit"))
        itemQuit.onAction = EventHandler { event: ActionEvent? -> exitProcess(0) }

        menu.items.add(itemQuit)

        return menu
    }

    private val menuEdit: Menu
        get() {
            val menu =
                Menu(Locale.getText("menubar.edit"))

            return menu
        }

    private val menuHelp: Menu
        get() {
            val menu =
                Menu(Locale.getText("menubar.help"))

            val itemAbout = MenuItem(
                getText(
                    "menubar.help.about",
                    Editor.Companion.TITLE
                )
            )
            itemAbout.onAction = EventHandler { event: ActionEvent? -> AboutWindow() }
            menu.items.add(itemAbout)

            val itemGithub =
                MenuItem(Locale.getText("menubar.help.repo"))
            itemGithub.onAction = EventHandler { event: ActionEvent? ->
                if (Desktop.isDesktopSupported() && Desktop.getDesktop()
                        .isSupported(Desktop.Action.BROWSE)
                ) {
                    try {
                        Desktop.getDesktop().browse(URI("https://github.com/Turboman3000/TurboEdit"))
                    } catch (e: IOException) {
                        logger.error("{} = {}", e.javaClass.getName(), e.message)
                    } catch (e: URISyntaxException) {
                        logger.error("{} = {}", e.javaClass.getName(), e.message)
                    }
                }
            }
            menu.items.add(itemGithub)

            return menu
        }
}
